src/Perl6/Metamodel/ContainerSpecProtocol.nqp

```diff
@@ -0,0 +1,23 @@
+role Perl6::Metamodel::ContainerSpecProtocol {
+    has $!code_pair;
+
+    method get_container_spec($obj) {
+        $!code_pair
+    }
+
+    method set_container_spec($obj, $code_pair) {
+        $!code_pair := $code_pair;
+    }
+
+    method publish_container_spec($obj) {
+        for self.mro($obj) -> $class {
+            if nqp::can($class.HOW, 'get_container_spec') {
+                my $code_pair := $class.HOW.get_container_spec($class);
+                if $code_pair {
```

Comment:
This is just idle curiosity, but could `if $class.HOW.get_container_spec($class) -> $code_pair {` be done? Does the `if ... ->` form work for/with binding?

Reply:
I get compilation errors if I do it like that.  I have no idea why and have no intent to look into NQP's guts to find out why that is not working.  I assume the `.HOW` in the expression is somehow throwing things off.